### PR TITLE
Average and biggest message added

### PR DESCRIPTION
**Pull request by ArgaAlan**

### CODE REVIEW

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 1]
```diff
@@ -1,12 +1,15 @@
+
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: remove unnecessary change

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 2]
```diff
@@ -54,15 +54,14 @@ public void storeMessage(Message message) {
...
-  	
-    Query query =
-        new Query("Message")
-            .setFilter(new Query.FilterPredicate("user", FilterOperator.EQUAL, user))
-            .addSort("timestamp", SortDirection.DESCENDING);
-    PreparedQuery results = datastore.prepare(query);
-
-   return convertEntitiesToMessages(results);
-  }
+
```
Comment (acolwell) [inline src/main/java/com/google/codeu/data/Datastore.java]: It looks like this doesn't actually change any functionality and just modifies the whitespace. Please revert this back to the old version so the whitespace matches the rest of the file. I think you may need to update your editor to use spaces instead of tabs to avoid this in the future.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 3]
```diff
@@ -38,4 +40,62 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
     response.getOutputStream().println(jsonObject.toString());
   }
-}
\ No newline at end of file
+  
+  /*
+   * Obtaining the average message by going through all the list
+   */
+  public void avgMessage(HttpServletRequest request, HttpServletResponse response) 
+		  throws IOException{
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: Please reformat this code to use spaces instead of tabs like the rest of the code and make it match indentation of the other code. The key think here is that it shouldn't be possible to notice the difference between code you added and the other developers. It also looks like you should add spaces in various if & for statements. The original code primarily follows Google Java style guide (https://google.github.io/styleguide/javaguide.html)

File: src/main/webapp/stats.html [hunk 4]
```diff
@@ -1,41 +1,50 @@
-<!DOCTYPE html>
```
Comment (acolwell) [inline src/main/webapp/stats.html:1]: I'm confused about why this appears to be showing the entire file changing instead of just showing the changes you have made. Are you making changes on 2 different computers that are using different line endings? Please make sure every machine you are working on use unix style line endings and not DOS style.

File: src/main/webapp/stats.html [hunk 5]
```diff
@@ -1,41 +1,50 @@
...
+<head>
+  <title>Stats</title>
+  <link rel="stylesheet" href="/css/main.css">
+
+  <script>
+    // Fetch stats and display them in the page.
+    function fetchStats(){
+      const url = '/stats';
+      fetch(url).then((response) => {
+        console.log(response);
```
Comment (acolwell) [inline src/main/webapp/stats.html]: nit: Please remove debugging code.

File: src/main/webapp/stats.html [hunk 6]
```diff
@@ -1,41 +1,50 @@
...
+        const statsContainer = document.getElementById('stats-container');
+        statsContainer.innerHTML = '';
+
+        const messageCountElement = buildStatElement('Message count: ' + stats.messageCount);
+        statsContainer.appendChild(messageCountElement);
+
+        const messageAverageElement = buildStatElement('Message average: ' + stats.messageAvg);
+        statsContainer.appendChild(messageAverageElement);
+
+        console.log(stats);
```
Comment (acolwell) [inline src/main/webapp/stats.html]: nit: remove these debugging statements as well.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 7]
```diff
@@ -38,4 +40,62 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     jsonObject.addProperty("messageCount", messageCount);
     response.getOutputStream().println(jsonObject.toString());
   }
-}
\ No newline at end of file
+  
+  /*
+   * Obtaining the average message by going through all the list
+   */
+  public void avgMessage(HttpServletRequest request, HttpServletResponse response)
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: These aren't going to get called by JavaScript. doGet() is what is called. You can change this functions signature to return an int and return the average. Then you can call it from doGet() and add a property to the jsonObject in that method. You can do the same thing for the biggestMessage() method.

**Review by acolwell — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 8]
```diff
@@ -1,41 +1,128 @@
...
+
+  /**
+   * Responds with site statistics in JSON.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    response.setContentType("application/json");
+    
+    com.google.codeu.data.Message message = new com.google.codeu.data.Message("Alan", "Ola k ase");
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: Please remove these 3 fake messages since the appear to only be used for your testing purposes.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 9]
```diff
@@ -1,41 +1,128 @@
...
+    //Number of messages property added to the Json
+    jsonObject.addProperty("messageCount", messageCount);
+    
+    //Store of the biggest message
+    com.google.codeu.data.Message biggest = allmessages.get(0);
+	
+    //Checking the size of the list
+	if(allmessages.size()==0) {
+		
+		jsonObject.addProperty("messageAvg", 0);
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: I'd suggest creating local variables for each of these computed values and then just move the code that assigns these local variables to properties at the bottom of this function. This has a few benefits over what you have here.

1. It makes sure all properties are assigned no matter the number of messages.
2. It allows you to use the Java type system to enforce what the properties should be. For example here you have the "biggestMessage" property set with an integer in one case and a string in another. If had a biggestMessage integer or String local variable, the compiler would have caught your inconsistent setting of this property.
3. It makes adding new properties less error prone because you only have to update the jsonObject in a single location at the bottom.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 10]
```diff
@@ -1,41 +1,128 @@
...
+			//count for the avg of all messages
+			countM+= allmessages.get(i).getText().length();
+			
+			//getting the biggest message
+			if(allmessages.get(i).getText().length()>biggest.getText().length()) {
+				biggest = allmessages.get(i);  
+			}
+			
+			//getting and saving the user and message count
+			if(!users.contains(allmessages.get(i).getUser())) {
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: This will be worst case O(M*N) complexity with an ArrayList. You should change this to use HashSet instead to get this down to O(M)

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 11]
```diff
@@ -1,41 +1,128 @@
...
+				users.add(allmessages.get(i).getUser());
+				usersNumberOfMessages.add(1);
+			} else {
+				usersNumberOfMessages.set(users.indexOf(allmessages.get(i).getUser()), usersNumberOfMessages.get(users.indexOf(allmessages.get(i).getUser()))+1);
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: This is worst case O(2N) because you are having to scan the users list twice here. If you change usersNumberOfMessages to be a HashMap<String,Integer> you can make this O(1) and I think the code will be easier to read.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 12]
```diff
@@ -1,41 +1,128 @@
...
+			}
+		}	  
+		
+		//Checking who has the most sent messages
+		int indexOfMoreMessages = -1;
+		int mostMessages = -1;
+		
+		for(int j = 0;j<usersNumberOfMessages.size();j++) {
+			if(usersNumberOfMessages.get(j)>mostMessages) {
+				indexOfMoreMessages = j;
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: I believe you can combine this with the logic in the loop above. When you update the count in usersNumberOfMessages you can also immediately check to see if it is the new maximum value.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 13]
```diff
@@ -72,40 +72,41 @@ public void storeMessage(Message message) {
    * @return a list of messages posted by all users, or empty list if no user has ever posted a
    *     message. List is sorted by time descending.
    */
-public List<Message> getAllMessages(){
+  public List<Message> getAllMessages(){
```
Comment (acolwell) [inline src/main/java/com/google/codeu/data/Datastore.java]: Please don't change the indentation of code you haven't made changes to and try to follow the style guide. If you happen to be using IntelliJ or Eclipse for your development here are some instructions on how to install plugins to help keep the code conforming to the style guides.

http://www.practicesofmastery.com/post/eclipse-google-java-style-guide/
http://www.practicesofmastery.com/post/intellij-checkstyle-google-java-style-guide/

**Review by acolwell — CHANGES_REQUESTED** [review]
In general these changes look pretty good. My main comments are around improving the runtime complexity of some of your calculations.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java (hunk 9 above)
Comment (nbaldy) [reply src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: @ArgaAlan I'm not completely sure that I follow these changes and suggestions; did you chose to implement this in your program? If so, could you possibly point me to where?

Comment (ArgaAlan) [reply src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: > @ArgaAlan I'm not completely sure that I follow these changes and suggestions; did you chose to implement this in your program? If so, could you possibly point me to where?

I did, before I first checked if the list was empty but I finally change it to compute everything and just add it to the JSON

Comment (nbaldy) [reply src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: oooh, so basically, you computed the variables, and then added them all to the json after?

**Comment by nbaldy** [comment]
So functionality seems good from local testing and from what I can understand from the code and talking to @ArgaAlan, changes have been implemented, aside from one whitespace change it seems good to me! However, I'd like to get final A-OK from @acolwell

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 14]
```diff
@@ -52,98 +50,102 @@ public void storeMessage(Message message) {
    * Gets messages posted by a specific user.
    *
    * @return a list of messages posted by the user, or empty list if user has never posted a
-   *     message. List is sorted by time descending.
-   *     Messages are public and will display on recipient's user page.
+   *         message. List is sorted by time descending. Messages are public and will display on
+   *         recipient's user page.
    */
+<<<<<<< HEAD
```
Comment (acolwell) [inline src/main/java/com/google/codeu/data/Datastore.java]: Please fix this merge conflict.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 15]
```diff
@@ -1,21 +1,22 @@
 /*
  * Copyright 2019 Google Inc.
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
```
Comment (acolwell) [inline src/main/java/com/google/codeu/data/Datastore.java:4]: Please put back the old license text formatting.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 16]
```diff
@@ -1,41 +1,100 @@
...
+      countM += allmessages.get(i).getText().length();
+
+      // getting the biggest message
+      if (allmessages.get(i).getText().length() > biggest.getText().length()) {
+        biggest = allmessages.get(i);
+      }
+
+
+      // getting and saving the user and message count and checking who has the most sent messages
+      if (!usersAndNumOfMessages.containsKey(allmessages.get(i).getUser())) {
```
Comment (acolwell) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java:69]: nit: Consider using a local variable to store allmessage.get(i).getUser() here so that you can use it here and everywhere below. I think that would make the code slightly easier to follow.

**Review by acolwell — APPROVED** [review]
LGTM overall. Feel free to merge once you've addressed the comments.
